lib/spack/llnl/util/tty/color.py

```diff
@@ -107,6 +107,56 @@ def __init__(self, message):
 _force_color = None
 
 
+def _enable_windows_vtp():
+    """Enable VTP in Windows consoles (CMD/PWSH/CONHOST)
+    Method based on the link below
+    https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#example-of-enabling-virtual-terminal-processing
+    """
+    import ctypes
+    import msvcrt
+    from ctypes import wintypes
+
+    ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x0004
+    DISABLE_NEWLINE_AUTO_RETURN = 0x0008
+    kernel32 = ctypes.WinDLL("kernel32")
+
+    def _err_check(result, func, args):
+        if not result:
+            raise ctypes.WinError(ctypes.get_last_error())
+        return args
+
+    kernel32.GetConsoleMode.errcheck = _err_check
+    kernel32.GetConsoleMode.argtypes = (
+        wintypes.HANDLE,  # hConsoleHandle, i.e. GetStdHandle output type
+        ctypes.POINTER(wintypes.DWORD),  # result of GetConsoleHandle
+    )
+    kernel32.SetConsoleMode.errcheck = _err_check
+    kernel32.SetConsoleMode.argtypes = (
+        wintypes.HANDLE,  # hConsoleHandle, i.e. GetStdHandle output type
+        wintypes.DWORD,  # result of GetConsoleHandle
+    )
+    # Use conout$ here to handle a redirectired stdout/get active console associated
+    # with spack
+    with open(r"\\.\CONOUT$", "w") as conout:
+        # Link above would use kernel32.GetStdHandle(-11) however this would not handle
+        # a redirected stdout appropriately, so we always refer to the current CONSOLE out
+        # which is defined as conout$ on Windows.
+        # linked example is follow more or less to the letter beyond this point
+        con_handle = msvcrt.get_osfhandle(conout.fileno())
+        dw_orig_mode = wintypes.DWORD()
+        kernel32.GetConsoleMode(con_handle, ctypes.byref(dw_orig_mode))
+        dw_new_mode_request = ENABLE_VIRTUAL_TERMINAL_PROCESSING | DISABLE_NEWLINE_AUTO_RETURN
+        dw_new_mode = dw_new_mode_request | dw_orig_mode.value
+        kernel32.SetConsoleMode(con_handle, wintypes.DWORD(dw_new_mode))
+
+
+# On Windows console handling for ASCI/VT100 sequences is not
+# on by default. Turn on before we try to write to console
+# with color
+if sys.platform == "win32":
+    _enable_windows_vtp()
```

Comment:
Would there be cases where this fails (and I assume one of the control commands in this function would raise an exception)? This could catch it, set `tty.color._force_color = False`, and then proceed (possibly also printing a debug message with the exception).

Reply:
You're right, ostensibly any of these commands could fail, and I have no idea how they'll behave on Windows 11, so I'll add the error handling.

Reply:
Addressed.